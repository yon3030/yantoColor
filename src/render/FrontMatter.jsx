import yaml from 'js-yaml';
import _ from 'lodash';

export function FrontMatter(props) {
  const { data } = props;

  if (data === undefined) {
    return '';
  } else if (_.isEmpty(data)) {
    return [
      '---\n',
      '---\n',
    ];
  } else {
    const frontMatter = yaml.safeDump(data);

    return [
      '---\n',
      frontMatter,
      '---\n',
    ];
  }
}
