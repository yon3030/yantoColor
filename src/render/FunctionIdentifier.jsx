import React from 'react';

export function FunctionIdentifier(props) {
  const { identifier, returnType } = props;

  const returnArrowAndType = returnType && [
    <span className="return-arrow">&rarr;</span>,
    <span className="return-type">returnType</span>,
  ];

  return <code className="function-identifier">{ identifier }{ returnArrowAndType }</code>;
}
