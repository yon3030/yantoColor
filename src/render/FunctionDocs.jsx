import React from 'react';

import { FunctionIdentifier } from './FunctionIdentifier';

export function FunctionDocs(props) {
  const { identifier, devdoc, contractName } = props;

  return (
    <>
    { '\n\n' }
    <h4>
      <a className="anchor" aria-hidden="true" id={ contractName + '.' + identifier }></a>
      <FunctionIdentifier withReturn { ...props } />
    </h4>
    { '\n\n' }
    { devdoc }
    { '\n\n' }
    </>
  );
}
