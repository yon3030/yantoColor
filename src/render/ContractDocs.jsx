import React from 'react';

import { FunctionDocs } from './FunctionDocs';
import { FunctionIdentifier } from './FunctionIdentifier';

export function ContractDocs(props) {
  const { name, devdoc, functions, events } = props;

  const functionDocs = functions.map(fn => (
    <FunctionDocs key={ fn.identifier } {...fn} contractName={ name } />
  ));

  const functionIndex = functions.length > 0 && <>
    <span className="contract-index-title">Functions</span>
    <ul>
      {
        functions.map(fn => (
          <li key={ fn.identifier }>
            <a href={ '#' + name + '.' + fn.identifier }>
              <FunctionIdentifier { ...fn } />
            </a>
          </li>
        ))
      }
    </ul>
  </>;

  const eventDocs = events.map(fn => (
    <FunctionDocs key={ fn.identifier } {...fn} contractName={ name } />
  ));

  const eventIndex = events.length > 0 && <>
    <span className="contract-index-title">Events</span>
    <ul>
      { 
        events.map(fn => (
          <li key={ fn.identifier }>
            <a href={ '#' + name + '.' + fn.identifier }>
              <FunctionIdentifier { ...fn } />
            </a>
          </li>
        ))
      }
    </ul>
  </>;

  return (
    <>
    { '\n\n' }
    ### `{ name }`
    { '\n\n' }
    { devdoc }
    { '\n\n' }
    <div className="contract-index">
      { functionIndex }
      { eventIndex }
    </div>
    { '\n\n' }
    { functionDocs }
    { '\n\n' }
    { eventDocs }
    { '\n\n' }
    </>
  );
}
