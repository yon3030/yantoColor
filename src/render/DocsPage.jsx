import React from 'react';

import { FrontMatter } from './FrontMatter';
import { ContractDocs } from './ContractDocs';

export function DocsPage(props) {
  const { frontMatter, title, intro, sections } = props;

  return [
    <FrontMatter data={ frontMatter } />,
    <h1>{ title }</h1>,
    '\n\n',
    intro,
    '\n\n',
    sections.map(section => {
      if (section.type === 'contracts') {
        return (
          <section>
            <h1>{ section.title }</h1>
            {
              section.contracts.map(contract => (
                <ContractDocs key={ contract.name } { ...contract } />
              ))
            }
          </section>
        );
      } else if (section.type === 'subdirectory') {
        return (
          <section>
            <DocsPage { ...section } />
          </section>
        );
      }
    }),
  ];
}

