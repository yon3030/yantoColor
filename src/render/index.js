import React from 'react';
import ReactDOMServer from 'react-dom/server';
import yaml from 'js-yaml';
import _ from 'lodash';

import { ContractDocs } from './ContractDocs';

export function renderDocs(docs) {
  const { frontMatter, intro, sections } = docs;

  const parts = [
    renderFrontMatter(frontMatter),
    intro,
    '<div class="contracts">',
    sections.map(renderSection),
    '</div>',
  ];

  return _.flattenDeep(parts)
    .filter(p => !_.isEmpty(p))
    .reduce(joinWithParagraphBreak);
}

export function renderSection(section) {
  const { title, contracts } = section;

  return [
    `## ${title}`,
    contracts.map(renderContract),
  ];
}

export function renderContract(contract) {
  return ReactDOMServer.renderToStaticMarkup(
    <ContractDocs { ...contract } />,
  );
}

export function renderFrontMatter(data) {
  if (_.isEmpty(data)) {
    return '---\n---\n';
  } else {
    const frontMatter = yaml.safeDump(data);
    return `---\n${frontMatter}---\n`;
  }
}

export function joinWithParagraphBreak(acc, str) {
  return acc.replace(/\n?\n?$/, str.replace(/^\n?\n?/, '\n\n'));
}
