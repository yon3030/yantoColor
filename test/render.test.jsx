import React from 'react';
import renderer from 'react-test-renderer';

import { FunctionDocs } from '../src/render/FunctionDocs';
import { ContractDocs } from '../src/render/ContractDocs';

describe('ContractDocs', function () {
  test('no functions', function () {
    const contract = {
      name: 'Foo',
      devdoc: 'devdoc',
      functions: [],
    };

    const tree = renderer
      .create(<ContractDocs { ...contract } />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('1 function', function () {
    const foo = {
      methodIdentifier: 'foo()',
      astNode: {
        nodeType: 'FunctionDefinition',
        name: 'foo',
        parameters: [],
      },
      devdoc: 'foo.devdoc',
    };

    const contract = {
      name: 'Foo',
      devdoc: 'devdoc',
      functions: [ foo ],
    };

    const tree = renderer
      .create(<ContractDocs { ...contract } />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('FunctionDocs', function () {
  test('basic function', function () {
    const foo = {
      methodIdentifier: 'foo()',
      astNode: {
        nodeType: 'FunctionDefinition',
        name: 'foo',
        parameters: [],
      },
      devdoc: 'foo.devdoc',
    };

    const tree = renderer
      .create(<FunctionDocs { ...foo } />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
