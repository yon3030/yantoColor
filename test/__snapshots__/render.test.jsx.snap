// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractDocs 1 function 1`] = `
Array [
  <h2>
    Foo
  </h2>,
  "devdoc",
  <h2>
    foo()
  </h2>,
  "foo.devdoc",
]
`;

exports[`ContractDocs no functions 1`] = `
Array [
  <h2>
    Foo
  </h2>,
  "devdoc",
]
`;

exports[`FunctionDocs basic function 1`] = `
Array [
  <h2>
    foo()
  </h2>,
  "foo.devdoc",
]
`;
